framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java

```diff
@@ -580,26 +636,79 @@ private void handleDimensions(
             if (AnnotationUtils.areSameIgnoringValues(dimType, BOTTOMVAL)) {
                 type.replaceAnnotation(BOTTOMVAL);
             } else {
-                List<Long> longLengths = null;
+                RangeOrListOfValues rolv = null;
                 if (AnnotationUtils.areSameByClass(dimType, IntRange.class)) {
-                    longLengths =
-                            ValueCheckerUtils.getValuesFromRange(getIntRange(dimType), Long.class);
+                    rolv = new RangeOrListOfValues(getRange(dimType));
                 } else if (AnnotationUtils.areSameByClass(dimType, IntVal.class)) {
-                    longLengths = getIntValues(dimType);
+                    rolv = new RangeOrListOfValues();
+                    rolv.addAll(rolv.convertLongsToInts(getIntValues(dimType)));
                 }
-                if (longLengths != null) {
-                    HashSet<Integer> lengths = new HashSet<Integer>(longLengths.size());
-                    for (Long l : longLengths) {
-                        lengths.add(l.intValue());
+                if (rolv != null) {
+                    AnnotationMirror newQual;
+                    if (rolv.isRange) {
+                        newQual = createArrayLenRangeAnnotation(rolv.range);
+                    } else {
+                        newQual = createArrayLenAnnotation(rolv.values);
                     }
-                    AnnotationMirror newQual = createArrayLenAnnotation(new ArrayList<>(lengths));
                     type.replaceAnnotation(newQual);
                 }
             }
         }
 
+        // An abstraction that can be either a range or a list of values that could come from an
+        // ArrayLen or IntVal. This abstraction reduces the number of cases that handleInitializers
+        // and handleDimensions must handle.
+        private class RangeOrListOfValues {
+            Range range;
+            List<Integer> values;
+            boolean isRange;
+
+            public RangeOrListOfValues(Integer... values) {
+                this.values = new ArrayList<>(values.length);
+                for (Integer l : values) {
+                    if (!this.values.contains(l)) {
+                        this.values.add(l);
+                    }
+                }
+                isRange = false;
+            }
+
+            public RangeOrListOfValues(Range range) {
+                this.range = range;
+                isRange = true;
+            }
+
+            public void addAll(List<Integer> newValues) {
+                for (Integer i : newValues) {
```

Comment:
This method does not account for the possibility that that this is a range not a list.